node/src/reactor/main_reactor/keep_up.rs

```diff
@@ -300,6 +301,44 @@ impl MainReactor {
                     parent_hash,
                     era_id,
                 } => {
+                    if era_id.is_genesis() && !self.validator_matrix.has_era(&era_id) {
+                        // We're syncing back and we're already in the genesis era. This means
+                        // that SyncLeap might not be able to provide us with the correct validator
+                        // set in case the chain has been started with pre 1.5 binary and the node
+                        // was just restarted. We might try to read the switch block of era 0 from
+                        // our local storage and populate the validator matrix.
+                        debug!("KeepUp: SyncBackInstruction: trying to saturate the validators for {era_id} from storage");
+                        match self
+                            .storage
+                            .transactional_get_switch_block_header_by_era_id(era_id)
```

Comment:
Not sure this solves it. If we already synced the first block of era 1 and now we are syncing the switch block of era 0 but we don't have the header yet, we still won't get the era validators from storage and sync leap won't help us. 
Should we maybe use `effect_builder.get_era_validators_from_contract_runtime()` instead maybe?

Reply:
Fair point, if the restart happens at the exact time switch block is being synced, we won't be able to recover. I need to dig into `get_era_validators_from_contract_runtime()`, but from a quick look it seems that we might still get the `GetEraValidatorsError::RootNotFound` and get stuck. The alternate solution with extended sync leap may turn out to be the most robust.

Reply:
As discussed, the change is implemented in https://github.com/casper-network/casper-node/pull/4116/commits/233f20f99e802d1368e5dd80ccbbff93e4749c4c